plugin/openassetio_manager_bal/BasicAssetLibraryInterface.py

```diff
@@ -233,6 +234,41 @@ def managementPolicy(self, traitSets, access, context, hostSession):
     def isEntityReferenceString(self, someString, hostSession):
         return someString.startswith(self.__entity_refrence_prefix())
 
+    @simulated_delay
+    def defaultEntityReference(
+        self, traitSets, defaultEntityAccess, context, hostSession, successCallback, errorCallback
+    ):
+        if not self.hasCapability(self.Capability.kDefaultEntityReferences):
+            super().defaultEntityReference(
+                traitSets,
+                defaultEntityAccess,
+                context,
+                hostSession,
+                successCallback,
+                errorCallback,
+            )
+            return
+
+        for idx, trait_set in enumerate(traitSets):
+            try:
+                entity_name = bal.default_entity(
+                    trait_set, kAccessNames[defaultEntityAccess], self.__library
+                )
+                entity_ref = None
+                # Entity can legitimately be None, meaning query was OK
+                # but there is no suitable default.
+                if entity_name is not None:
+                    entity_ref = self.__build_entity_ref(
+                        bal.EntityInfo(
+                            name=entity_name,
+                            access=kAccessNames[defaultEntityAccess],
+                            version=None,
+                        )
+                    )
+                successCallback(idx, entity_ref)
```

Comment:
As I'm not sure of how tightly coupled entity_name and entity_ref are (are they always both non-None, always both None, or one can be None while the other isn't?), is there a possibility here that `successCallback` is called with entity_ref=None?

And does that need to be checked for, as 'success' and None don't necessarily go hand in hand IMO.

Although I'm not 100% sure on this, because I see None is mentioned in the commit message as a possible value.

Reply:
`entity_ref` is essentially `"bal:///" + entity_name` (then potentially a `?v=X` version suffix, but that's not relevant here).

`None` is indeed a legit value. In C++ the corresponding type is a `std::optional`. `None` means that the query was well formed but there is no appropriate default reference. We didn't want to make that an error case because it's a perfectly legitimate response from a manager.

There is a unit test for the `None` case (`test_when_no_default_then_entity_ref_is_None`).